pkg/api/handlers/libpod/artifacts.go

```
@@ -0,0 +1,290 @@
+	if err != nil {
+		if errors.Is(err, libartifact_types.ErrArtifactNotExist) {
+			utils.ArtifactNotFound(w, name, err)
+			return
+		} else {
+			utils.InternalServerError(w, err)
+			return
+		}
+	}
+
+	utils.WriteResponse(w, http.StatusOK, report)
+}
+
+func ListArtifact(w http.ResponseWriter, r *http.Request) {
+	runtime := r.Context().Value(api.RuntimeKey).(*libpod.Runtime)
+
+	imageEngine := abi.ImageEngine{Libpod: runtime}
+
+	artifacts, err := imageEngine.ArtifactList(r.Context(), entities.ArtifactListOptions{})
+	if err != nil {
+		utils.InternalServerError(w, err)
+		return
+	}
+
+	utils.WriteResponse(w, http.StatusOK, artifacts)
+}
+
+func PullArtifact(w http.ResponseWriter, r *http.Request) {
+	runtime := r.Context().Value(api.RuntimeKey).(*libpod.Runtime)
+	decoder := r.Context().Value(api.DecoderKey).(*schema.Decoder)
+
+	query := struct {
+		Name       string             `schema:"name"`
+		Retry      uint               `schema:"retry"`
+		RetryDelay string             `schema:"retryDelay"`
+		TLSVerify  types.OptionalBool `schema:"tlsVerify"`
+	}{}
+
+	if err := decoder.Decode(&query, r.URL.Query()); err != nil {
+		utils.Error(w, http.StatusBadRequest, fmt.Errorf("failed to parse parameters for %s: %w", r.URL.String(), err))
+		return
+	}
+
+	if query.Name == "" {
+		utils.Error(w, http.StatusBadRequest, errors.New("name parameter is required"))
+		return
+	}
+
+	artifactsPullOptions := entities.ArtifactPullOptions{
+		InsecureSkipTLSVerify: query.TLSVerify,
+	}
+
+	// If TLS verification is explicitly specified (True or False) in the query,
+	// set the InsecureSkipTLSVerify option in the artifactsPullOptions accordingly.
+	// If TLSVerify was not set in the query, OptionalBoolUndefined is used and
+	// handled later based off the target registry configuration.
+	if s := query.TLSVerify; s != types.OptionalBoolUndefined {
+		if s == types.OptionalBoolTrue {
+			artifactsPullOptions.InsecureSkipTLSVerify = types.NewOptionalBool(false)
+		} else {
```

(Security systems should “fail closed”; if this ever encountered an invalid `types.OptionalBool` value, it should either fail, or set `InsecureSkip…` to `false`, not `true`.

Arguably, most pre-existing code also gets this wrong.

Really, the conversions between the `SkipTLSVerify` values in Go, and the `TLSVerify` values in the HTTP API, should all be centralized in a single utility. But that’s not this PR.)

---

I'm not sure what I can do here, I agree it's a really inconvenient conversion. I would be nice to just set the query parameter to SkipTLSVerify because currently the OptionalBool logic fails the wrong way.

---

Longer-term,
```go
func someSharedPackage.insecureTLSVerifyFromPositiveTLSVerify(in types.OptionalBool) types.OptionalBool {
  switch in {
  case types.OptionalBoolUndefined:
    return types.OptionalBoolUndefined
  case types.OptionalBoolFalse: return …
  case types.OptionalBoolTrue: return …
  default: // Should never happen, require TLS to be conservative
    return types.NewOptionalBoolFalse
  }
}
```
+, if I could just wish for things, a comprehensive unit test for that helper